File: src/harness_core/dispatch/plugin_system.py
"""Phase 7: PluginSystem — load, validate, and enforce permissions for plugins."""

from __future__ import annotations

import json
import threading
import time
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Any


PLUGIN_SYSTEM_SCHEMA_VERSION = "plugin_system.v1"

TRUST_LEVEL_COMMUNITY = "community"
TRUST_LEVEL_VERIFIED = "verified"
TRUST_LEVEL_OFFICIAL = "official"

VALID_TRUST_LEVELS = (TRUST_LEVEL_COMMUNITY, TRUST_LEVEL_VERIFIED, TRUST_LEVEL_OFFICIAL)

TRUST_PERMISSIONS = {
    TRUST_LEVEL_COMMUNITY: frozenset({"dispatch:read"}),
    TRUST_LEVEL_VERIFIED: frozenset({"dispatch:read", "dispatch:write"}),
    # OFFICIAL: empty set means unrestricted — check is bypassed in _trust_permission_allowed
    TRUST_LEVEL_OFFICIAL: frozenset(),
}

ALL_KNOWN_PERMISSIONS = frozenset({
    "dispatch:read",
    "dispatch:write",
    "provider:execute",
    "config:read",
    "config:write",
    "ledger:read",
    "ledger:write",
})


class TrustLevel(Enum):
    COMMUNITY = TRUST_LEVEL_COMMUNITY
    VERIFIED = TRUST_LEVEL_VERIFIED
    OFFICIAL = TRUST_LEVEL_OFFICIAL


REQUIRED_MANIFEST_FIELDS = (
    "schema_version",
    "plugin_id",
    "name",
    "version",
    "author",
    "permissions",
    "entrypoints",
    "compatible_dispatcher_versions",
    "required_env",
    "network_access",
    "filesystem_access",
    "trust_level",
)


@dataclass(frozen=True)
class PluginManifest:
    plugin_id: str
    name: str
    version: str
    author: str
    permissions: tuple[str, ...] = ()
    entrypoints: tuple[str, ...] = ()
    compatible_dispatcher_versions: tuple[str, ...] = ()
    required_env: tuple[str, ...] = ()
    network_access: bool = False
    filesystem_access: bool = False
    signature: str | None = None
    trust_level: str = TRUST_LEVEL_COMMUNITY
    schema_version: str = "plugin_manifest.v1"

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "plugin_id": self.plugin_id,
            "name": self.name,
            "version": self.version,
            "author": self.author,
            "permissions": list(self.permissions),
            "entrypoints": list(self.entrypoints),
            "compatible_dispatcher_versions": list(self.compatible_dispatcher_versions),
            "required_env": list(self.required_env),
            "network_access": self.network_access,
            "filesystem_access": self.filesystem_access,
            "signature": self.signature,
            "trust_level": self.trust_level,
        }


@dataclass(frozen=True)
class LoadedPlugin:
    manifest: PluginManifest
    module_name: str
    loaded_at: float = field(default_factory=time.time)
    status: str = "loaded"


class PluginSystem:
    """Loads plugins from JSON manifests and enforces trust-level permissions."""

    def __init__(self) -> None:
        self._plugins: dict[str, LoadedPlugin] = {}
        self._lock = threading.RLock()

    def load_plugin(
        self,
        manifest_path: str | Path,
        plugin_dir: str | Path,
    ) -> LoadedPlugin:
        manifest_path = Path(manifest_path)
        plugin_dir = Path(plugin_dir)

        if not manifest_path.exists():
            raise FileNotFoundError(f"Manifest not found: {manifest_path}")
        if not plugin_dir.exists():
            raise NotADirectoryError(f"Plugin directory not found: {plugin_dir}")

        with open(manifest_path, "r") as f:
            raw = json.load(f)

        manifest = _parse_manifest(raw)
        errors = _validate_manifest_fields(manifest)
        if errors:
            raise ValueError(f"Invalid manifest: {'; '.join(errors)}")

        if not _trust_permission_allowed(manifest.trust_level, manifest.permissions):
            raise PermissionError(
                f"Trust level '{manifest.trust_level}' not allowed for permissions: "
                f"{[p for p in manifest.permissions if p not in TRUST_PERMISSIONS.get(manifest.trust_level, frozenset())]}"
            )

        module_name = f"harness_plugin.{manifest.plugin_id}"

        with self._lock:
            if manifest.plugin_id in self._plugins:
                del self._plugins[manifest.plugin_id]

            loaded = LoadedPlugin(manifest=manifest, module_name=module_name)
            self._plugins[manifest.plugin_id] = loaded
            return loaded

    def unload_plugin(self, plugin_id: str) -> bool:
        with self._lock:
            if plugin_id in self._plugins:
                del self._plugins[plugin_id]
                return True
            return False

    def check_permission(self, plugin_id: str, permission: str) -> bool:
        with self._lock:
            loaded = self._plugins.get(plugin_id)
        if loaded is None:
            return False
        if permission not in ALL_KNOWN_PERMISSIONS:
            return False
        return permission in loaded.manifest.permissions

    def list_plugins(self) -> list[LoadedPlugin]:
        with self._lock:
            return list(self._plugins.values())

    def get_plugin(self, plugin_id: str) -> LoadedPlugin | None:
        with self._lock:
            return self._plugins.get(plugin_id)


def _parse_manifest(raw: dict[str, Any]) -> PluginManifest:
    perms = raw.get("permissions", [])
    if isinstance(perms, list):
        perms = tuple(str(p) for p in perms)
    else:
        perms = ()

    entrypoints = raw.get("entrypoints", [])
    if isinstance(entrypoints, list):
        entrypoints = tuple(str(e) for e in entrypoints)
    else:
        entrypoints = ()

    compat = raw.get("compatible_dispatcher_versions", [])
    if isinstance(compat, list):
        compat = tuple(str(c) for c in compat)
    else:
        compat = ()

    req_env = raw.get("required_env", [])
    if isinstance(req_env, list):
        req_env = tuple(str(r) for r in req_env)
    else:
        req_env = ()

    return PluginManifest(
        plugin_id=str(raw.get("plugin_id", "")),
        name=str(raw.get("name", "")),
        version=str(raw.get("version", "")),
        author=str(raw.get("author", "")),
        permissions=perms,
        entrypoints=entrypoints,
        compatible_dispatcher_versions=compat,
        required_env=req_env,
        network_access=bool(raw.get("network_access", False)),
        filesystem_access=bool(raw.get("filesystem_access", False)),
        signature=raw.get("signature"),
        trust_level=str(raw.get("trust_level", TRUST_LEVEL_COMMUNITY)),
        schema_version=str(raw.get("schema_version", "plugin_manifest.v1")),
    )


def _validate_manifest_fields(manifest: PluginManifest) -> list[str]:
    errors: list[str] = []
    if not manifest.plugin_id:
        errors.append("plugin_id is required")
    if not manifest.name:
        errors.append("name is required")
    if not manifest.version:
        errors.append("version is required")
    if not manifest.author:
        errors.append("author is required")
    if manifest.trust_level not in VALID_TRUST_LEVELS:
        errors.append(f"trust_level must be one of {VALID_TRUST_LEVELS}, got '{manifest.trust_level}'")
    for perm in manifest.permissions:
        if perm not in ALL_KNOWN_PERMISSIONS:
            errors.append(f"unknown permission '{perm}'")
    return errors


def _trust_permission_allowed(trust_level: str, permissions: tuple[str, ...]) -> bool:
    if trust_level == TRUST_LEVEL_OFFICIAL:
        return True
    allowed = TRUST_PERMISSIONS.get(trust_level, frozenset())
    for perm in permissions:
        if perm not in allowed:
            return False
    return True

File: engine/src/infrastructure/plugin_system.rs
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};

pub const PLUGIN_SYSTEM_SCHEMA_VERSION: &str = "plugin_system.v1";

pub const TRUST_LEVEL_COMMUNITY: &str = "community";
pub const TRUST_LEVEL_VERIFIED: &str = "verified";
pub const TRUST_LEVEL_OFFICIAL: &str = "official";

pub const VALID_TRUST_LEVELS: &[&str] = &[
    TRUST_LEVEL_COMMUNITY,
    TRUST_LEVEL_VERIFIED,
    TRUST_LEVEL_OFFICIAL,
];

pub const ALL_KNOWN_PERMISSIONS: &[&str] = &[
    "dispatch:read",
    "dispatch:write",
    "provider:execute",
    "config:read",
    "config:write",
    "ledger:read",
    "ledger:write",
];

pub fn trust_permissions(trust_level: &str) -> HashSet<String> {
    match trust_level {
        TRUST_LEVEL_COMMUNITY => {
            let mut s = HashSet::new();
            s.insert("dispatch:read".to_string());
            s
        }
        TRUST_LEVEL_VERIFIED => {
            let mut s = HashSet::new();
            s.insert("dispatch:read".to_string());
            s.insert("dispatch:write".to_string());
            s
        }
        TRUST_LEVEL_OFFICIAL => HashSet::new(), // empty = unrestricted
        _ => HashSet::new(),
    }
}

pub const REQUIRED_MANIFEST_FIELDS: &[&str] = &[
    "schema_version",
    "plugin_id",
    "name",
    "version",
    "author",
    "permissions",
    "entrypoints",
    "compatible_dispatcher_versions",
    "required_env",
    "network_access",
    "filesystem_access",
    "trust_level",
];

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PluginManifest {
    pub schema_version: String,
    pub plugin_id: String,
    pub name: String,
    pub version: String,
    pub author: String,
    pub permissions: Vec<String>,
    pub entrypoints: Vec<String>,
    pub compatible_dispatcher_versions: Vec<String>,
    pub required_env: Vec<String>,
    pub network_access: bool,
    pub filesystem_access: bool,
    pub signature: Option<String>,
    pub trust_level: String,
}

impl PluginManifest {
    pub fn to_dict(&self) -> serde_json::Value {
        serde_json::to_value(self).unwrap_or(serde_json::Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct LoadedPlugin {
    pub manifest: PluginManifest,
    pub module_name: String,
    pub loaded_at: f64,
    pub status: String,
}

pub struct PluginSystem {
    plugins: HashMap<String, LoadedPlugin>,
}

impl Default for PluginSystem {
    fn default() -> Self {
        Self::new()
    }
}

impl PluginSystem {
    pub fn new() -> Self {
        Self {
            plugins: HashMap::new(),
        }
    }

    pub fn load_plugin_from_manifest(
        &mut self,
        manifest: PluginManifest,
        now: f64,
    ) -> Result<LoadedPlugin, String> {
        let errors = validate_manifest_fields(&manifest);
        if !errors.is_empty() {
            return Err(format!("Invalid manifest: {}", errors.join("; ")));
        }

        if !trust_permission_allowed(&manifest.trust_level, &manifest.permissions) {
            let denied: Vec<String> = manifest
                .permissions
                .iter()
                .filter(|p| {
                    let allowed = trust_permissions(&manifest.trust_level);
                    !allowed.contains(p.as_str())
                })
                .cloned()
                .collect();
            return Err(format!(
                "Trust level '{}' not allowed for permissions: {:?}",
                manifest.trust_level, denied
            ));
        }

        let module_name = format!("harness_plugin.{}", manifest.plugin_id);
        let loaded = LoadedPlugin {
            manifest: manifest.clone(),
            module_name,
            loaded_at: now,
            status: "loaded".to_string(),
        };
        self.plugins
            .insert(manifest.plugin_id.clone(), loaded.clone());
        Ok(loaded)
    }

    pub fn unload_plugin(&mut self, plugin_id: &str) -> bool {
        self.plugins.remove(plugin_id).is_some()
    }

    pub fn check_permission(&self, plugin_id: &str, permission: &str) -> bool {
        let loaded = match self.plugins.get(plugin_id) {
            Some(l) => l,
            None => return false,
        };
        if !ALL_KNOWN_PERMISSIONS.contains(&permission) {
            return false;
        }
        loaded
            .manifest
            .permissions
            .contains(&permission.to_string())
    }

    pub fn list_plugins(&self) -> Vec<&LoadedPlugin> {
        self.plugins.values().collect()
    }

    pub fn get_plugin(&self, plugin_id: &str) -> Option<&LoadedPlugin> {
        self.plugins.get(plugin_id)
    }
}

pub fn parse_manifest(raw: &serde_json::Value) -> Result<PluginManifest, String> {
    let get_str = |key: &str| -> String {
        raw.get(key)
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string()
    };
    let get_str_vec = |key: &str| -> Vec<String> {
        raw.get(key)
            .and_then(|v| v.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|v| v.as_str().map(String::from))
                    .collect()
            })
            .unwrap_or_default()
    };

    Ok(PluginManifest {
        schema_version: get_str("schema_version"),
        plugin_id: get_str("plugin_id"),
        name: get_str("name"),
        version: get_str("version"),
        author: get_str("author"),
        permissions: get_str_vec("permissions"),
        entrypoints: get_str_vec("entrypoints"),
        compatible_dispatcher_versions: get_str_vec("compatible_dispatcher_versions"),
        required_env: get_str_vec("required_env"),
        network_access: raw
            .get("network_access")
            .and_then(|v| v.as_bool())
            .unwrap_or(false),
        filesystem_access: raw
            .get("filesystem_access")
            .and_then(|v| v.as_bool())
            .unwrap_or(false),
        signature: raw
            .get("signature")
            .and_then(|v| v.as_str())
            .map(String::from),
        trust_level: get_str("trust_level"),
    })
}

pub fn validate_manifest_fields(manifest: &PluginManifest) -> Vec<String> {
    let mut errors = Vec::new();
    if manifest.plugin_id.is_empty() {
        errors.push("plugin_id is required".to_string());
    }
    if manifest.name.is_empty() {
        errors.push("name is required".to_string());
    }
    if manifest.version.is_empty() {
        errors.push("version is required".to_string());
    }
    if manifest.author.is_empty() {
        errors.push("author is required".to_string());
    }
    if !VALID_TRUST_LEVELS.contains(&manifest.trust_level.as_str()) {
        errors.push(format!(
            "trust_level must be one of {:?}, got '{}'",
            VALID_TRUST_LEVELS, manifest.trust_level
        ));
    }
    for perm in &manifest.permissions {
        if !ALL_KNOWN_PERMISSIONS.contains(&perm.as_str()) {
            errors.push(format!("unknown permission '{}'", perm));
        }
    }
    errors
}

pub fn trust_permission_allowed(trust_level: &str, permissions: &[String]) -> bool {
    if trust_level == TRUST_LEVEL_OFFICIAL {
        return true;
    }
    let allowed = trust_permissions(trust_level);
    permissions.iter().all(|p| allowed.contains(p.as_str()))
}
